### ISSUE TITLE
connect with infographiq

### ISSUE THREAD

**Issue by 7yl4r** [marinebon/fk-iea#4]
Good morning @bbest : 
I woke up with some thoughts about how infographiq can fit into the blogdown/hugo paradigm.
By wrapping infographiq as a hugo "theme" it will be easy to use while still staying separate from the site content.

The theme being used for this site right now (in [./themes](https://github.com/marinebon/fk-iea/tree/master/themes)) is "book", and you can look there for an example theme.

To deploy infographiq as I theme would require:
1. "layouts" `.html` for the main infographic view and modals
2. code to generate the main infograhpic page from the `.svg` file and `svg-links.csv`

(2) is going to be a little weird since hugo themes are typically restricted to `.html` & `.md`, but I _think_ bookdown will handle it if there is a `.Rmd` at `./layouts/home.Rmd`. :man_shrugging: . 

Worst case we'll use an R script to generate `/content/_index` for the individual site similar to how infographiq works currently.

That should work with the newer r2d3 stuff, right?

**Comment by bbest** [comment]
Hey @7yl4r,

I got basic infographiq functionality working with https://github.com/marinebon/infographiq/commit/63d06793ee2f9e7639c7006af0c28ecc8ebd7c15 by explicitly adding dependency to bootstrap, which was loading for default Rmarkdown page but not the blogdown site. 

Now am trying to figure out:
1. Why whole screen goes gray during modal window popup and width not properly sized. Perhaps because of "Failed to load resource" JS errors?
2. Why getting  "Failed to load resource" JS errors around `oxygen-v8-latin-*` when loading acidification modal, which are used in the left sidebar nav menu. Perhaps moot if wanting to remove the nav menu anyways.
3. How to not render the nav menu in the modals since obscures content. Tried adding front matter.
4. Why not loading regular size modal.

![image](https://user-images.githubusercontent.com/2837257/60543246-acec3980-9cca-11e9-98f1-0fde67e05300.png)

I think most of issues have to do with layout in this blogdown not being based off bootstrap.

**Comment by bbest** [comment]
Going to try porting content over to 

* [devcows/hugo-universal-theme](https://github.com/devcows/hugo-universal-theme); see [demo](https://themes.gohugo.io/theme/hugo-universal-theme/)

Since this theme is based on bootstrap with a nice look and is referenced in [1.6 Other themes | blogdown documentation](https://bookdown.org/yihui/blogdown/other-themes.html).

Will try to enable toggling of nav menu, aka toc (table of contents) via front matter for modals a la:

```yaml
output:
  blogdown::html_page:
    toc: true
```

per **Show the table of contents (TOC)** in [2.5 Templates | blogdown: Creating Websites with R Markdown](https://bookdown.org/yihui/blogdown/templates.html#how-to), although that used the [default theme](https://bookdown.org/yihui/blogdown/themes.html#the-default-theme) hugo-lithium without bootstrap.

Actually, probably best to have a different content type, aka layout / template / archetype.

TODO: check out [widgetframe](https://bhaskarvk.github.io/widgetframe/) R package to embed htmlwidgets in responsive iframes and not conflict with parent page's JS/CSS.
- use [`frameableWidget()`](https://bhaskarvk.github.io/widgetframe/#frameablewidget-function-) with [`saveWidget(selfcontained = F, libdir = ...)`](https://www.rdocumentation.org/packages/htmlwidgets/versions/1.3/topics/saveWidget) and [setup pym.js](http://blog.apps.npr.org/pym.js/) for site

Create new folder-based `archetypes/*.md` for page, modal, etc:
* [A Spoonful of Hugo: Archetypes | Alison Hill](https://alison.rbind.io/post/2019-02-19-hugo-archetypes/)
* [Archetypes | Hugo](https://gohugo.io/content-management/archetypes/)

Another very cool & powerful, blogdown based theme: [Academic](https://sourcethemes.com/academic/)

**Comment by bbest** [comment]
Still refactoring for fully working site

**Comment by bbest** [comment]
Strategies to get it to work, since realizing:

> So I figured out that the problem was not bootstrap w/ infographiq, but placement of modal in html page. When I moved to bottom, it works. Now just trying to figure out how to get modal properly injected, probably by replacing div tag with alternate configuration.

1. Inject as [widgetframe](https://cran.r-project.org/web/packages/widgetframe/vignettes/Using_widgetframe.html) so all the CSS and JSS for popping up the modal is independent of parent page containing the infographiq widget embedded as an iframe. 
    - In order to get this to work, I now have to use <code>htmlwidgets::<a href="https://www.rdocumentation.org/packages/htmlwidgets/versions/1.3/topics/createWidget)">createWidget</a>()</code> to create a proper widget, per [Package Development • r2d3](https://rstudio.github.io/r2d3/articles/package_development.html).
1. Provide an optional argument to <code><a href="https://marinebon.github.io/infographiq/reference/info_svg.html">info_svg</a>(div_id = "modalhere")</code> to reference a div the user places in a workable part of the html, such as just before `</body>` per [Considerations for Styling a Modal | CSS-Tricks](https://css-tricks.com/considerations-styling-modal/). Then the div tag would be replaced with the modal html by passing it in from info_svg as `options.html` (similar to `options.svg`) and handled by [`infographiq.js`](https://github.com/marinebon/infographiq/blob/master/inst/infographiq.js)

**Comment by bbest** [comment]
Hey @7yl4r ,

I think we're finally set with a solution for infographiq in the wild with these commits:

- [marinebon/**fk-iea**@59250f7: fully working info_svg(modal_html_provided=T) w/ partials/modal.html & `modal_id` hugo param](https://github.com/marinebon/fk-iea/commit/59250f7aefae6aeb83077e1e5dd1422087086833)
- [marinebon/**infographiq**@f8a8a3a: + arg `modal_html_provided=F` for `info_svg()` for funky layouts](https://github.com/marinebon/infographiq/commit/f8a8a3a53be234369a8483d5153179814882a75a)

Here's how it works now on this blogdown site:

1. For any posts that use a modal, include the following in the front matter:
    ```yaml
    modal_id: "modal"
    ```
    This injects the html needed for the modal nearest to the `</body>` tag outside of layout harm's way, thanks to these [custom blogdown layout](https://bookdown.org/yihui/blogdown/custom-layouts.html) elements:
    - detection of the post parameter here to inject the partial:
       https://github.com/marinebon/fk-iea/blob/59250f7aefae6aeb83077e1e5dd1422087086833/layouts/posts/baseof.html#L27-L29
    - injection of the output from [`infographiq::modal_html()`](https://marinebon.github.io/infographiq/reference/modal_html.html) here:
       https://github.com/marinebon/fk-iea/blob/59250f7aefae6aeb83077e1e5dd1422087086833/layouts/partials/modal.html#L1-L16
1. Use the argument `modal_html_provided=T` (and default `modal_id = "modal"`) with [infographiq::`info_svg()`](https://marinebon.github.io/infographiq/reference/info_svg.html) to use modal html already provided (vs injecting into funky place for layout within body):
https://github.com/marinebon/fk-iea/blob/59250f7aefae6aeb83077e1e5dd1422087086833/content/posts/externalities.Rmd#L42

**Comment by bbest** [comment]
fixed issues by including the following in `infographiq.js`:

```js
$('#'+ options.modal_id).appendTo('body');
```

in commit https://github.com/marinebon/infographiq/commit/aecb9fa652631e8914e6b944ab1da05e8eff4932, as suggested in [Modal backdrop displays over modal in fixed container · Issue #23916 · twbs/bootstrap](https://github.com/twbs/bootstrap/issues/23916#issuecomment-476794355)
